### PR TITLE
fs-1126-postgres-unit-testing

### PR DESCRIPTION
**Pull request by SentientHousePlant**

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -71,12 +80,47 @@ jobs:
...
+        uses: actions/checkout@v2
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: 3.10.1
+      - name: install dependencies
+        run: python -m pip install --upgrade pip && python -m pip install -r requirements-dev.txt
+      - name: run unit tests
+        # pytest -m "not accessibility" runs every test which is not marked 
+        # accessibility.
```
Comment (srh-sloan) [inline .github/workflows/deploy.yml:117]: Should we still be excluding accessibility tests here? If not pls remove the comment :)

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -71,12 +80,47 @@ jobs:
...
+        run: python -m pip install pytest && python -m pytest
   
   testing:
     runs-on: ubuntu-latest
     environment: Dev
-    needs: testing-unit
-    if: ${{ inputs.api == false }}
+    if: ${{ always() && (needs.testing-unit.result=='success' || needs.testing-unit-postgres.result=='success') && inputs.api == false}}
```
Comment (srh-sloan) [inline .github/workflows/deploy.yml:123]: Do we need the `always()` here?

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -98,7 +142,7 @@ jobs:
         run: python -m pip install pytest && python -m pytest -m "not accessibility" .
 
   deploy_dev:
-    if: ${{ always() && (needs.testing-unit.result=='success' || needs.testing.result=='success')}}
+    if: ${{ always() && (needs.testing-unit.result=='success' || needs.testing.result=='success' || needs.testing-unit-postgres.result=='success') && inputs.deploy_to_dev == true}}
```
Comment (srh-sloan) [inline .github/workflows/deploy.yml:145]: Same here about the `always()` - also if either of those jobs has failed, won't it skip later jobs anyway?

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -187,6 +224,29 @@ jobs:
...
+      - name: install dependencies
+        run: python -m pip install --upgrade pip && python -m pip install -r requirements.txt bandit safety
+      - name: Bandit
+        run: bandit -r ./app
+      - name: Safety
+        run: safety check
+      - name: ZAP Scan
+        uses: zaproxy/action-full-scan@v0.3.0
+        with:
+          target: 'https://${{inputs.app_name}}-dev.london.cloudapps.digital/'
```
Comment (srh-sloan) [inline .github/workflows/deploy.yml:248]: If this job needs `deploy_test` should we be targeting `test` here not `dev`?

**Review by srh-sloan — COMMENTED** [review]
A few comments but looks good!

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (SentientHousePlant) [reply .github/workflows/deploy.yml:141]: thats right, its the only way to have blocking jobs that dont make the job conditional

**Review by gio-karunakaran — APPROVED** [review]

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -71,12 +78,46 @@ jobs:
...
+    if: ${{ inputs.postgres_unit_testing == true }}
+    services:
+      # Label used to access the service container
+      postgres:
+        # Docker Hub image
+        image: postgres
+        # Provide the password for postgres
+        env:
+          POSTGRES_PASSWORD: postgres
+          POSTGRES_DB: fsd_app_store_test
```
Comment (gidsg) [inline .github/workflows/deploy.yml:94]: Can we not hard code this as it's app specific and make it an argument instead?

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -71,12 +82,46 @@ jobs:
         # pytest -m "not accessibility" runs every test which is not marked 
         # accessibility.
         run: python -m pip install pytest && python -m pytest -m "not accessibility" .
+        
+  testing-unit-postgres: 
+    runs-on: ubuntu-latest
+    environment: Dev
+    if: ${{ inputs.postgres_unit_testing == true }}
+    services:
```
Comment (gidsg) [inline .github/workflows/deploy.yml:90]: Is it possible to just create the service conditionally based on the flag so we don't repeat the python stuff between the steps?

Comment (SentientHousePlant) [reply .github/workflows/deploy.yml:90]: Whole jobs can be parameterised using matrices but from reading the documentation it seems that services cannot be made conditional.

**Review by gidsg — APPROVED** [review]
LGTM 👍  could you update the title and description to make it clearer which story this is linked to (i.e. the story which necessitated this change)?

**Pull request merged**